binary_transparency/firmware/internal/ftmap/log.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ftmap
+
+import (
+	"crypto"
+	"encoding/binary"
+	"fmt"
+	"reflect"
+	"sort"
+
+	"github.com/apache/beam/sdks/go/pkg/beam"
+
+	"github.com/google/trillian/experimental/batchmap"
+	"github.com/google/trillian/merkle/compact"
+	"github.com/google/trillian/merkle/coniks/hasher"
+
+	"golang.org/x/mod/sumdb/tlog"
+)
+
+func init() {
+	beam.RegisterFunction(makeDeviceReleaseLogFn)
+	beam.RegisterType(reflect.TypeOf((*moduleLogHashFn)(nil)).Elem())
+	beam.RegisterType(reflect.TypeOf((*DeviceReleaseLog)(nil)).Elem())
+}
+
+// DeviceReleaseLog represents firmware releases found for a single device ID.
+// Entries are ordered by their sequence in the original log.
+type DeviceReleaseLog struct {
+	DeviceID  string
+	Revisions []uint64
+}
+
+// MakeReleaseLogs takes all firmwareLogEntrys and processes these by
+// DeviceID in order to create logs of release revisions. The versions for
+// each DeviceID are sorted (by ID in the original log), and a log is
+// constructed for each device. This method returns two PCollections:
+// 1. the first is of type Entry; the key/value data to include in the map
+// 2. the second is of type DeviceReleaseLog.
+func MakeReleaseLogs(s beam.Scope, treeID int64, logEntries beam.PCollection) (beam.PCollection, beam.PCollection) {
+	keyed := beam.ParDo(s, func(l *firmwareLogEntry) (string, *firmwareLogEntry) { return l.Firmware.DeviceID, l }, logEntries)
+	logs := beam.ParDo(s, makeDeviceReleaseLogFn, beam.GroupByKey(s, keyed))
+	return beam.ParDo(s, &moduleLogHashFn{TreeID: treeID}, logs), logs
+}
+
+type moduleLogHashFn struct {
+	TreeID int64
+
+	rf *compact.RangeFactory
+}
+
+func (fn *moduleLogHashFn) Setup() {
+	fn.rf = &compact.RangeFactory{
+		Hash: func(left, right []byte) []byte {
+			// There is no particular need for using this hash function, but it was convenient.
```

Comment:
I wonder if it's worth just having our own defined locally to avoid confusion from the dep on golang sumdb internals?

Reply:
Yeah, good call.